File: java/matching-lean-flyweight/src/main/java/io/github/giovanicaprison/matching/lean/flyweight/LeanEngine.java
package io.github.giovanicaprison.matching.lean.flyweight;

import io.github.giovanicaprison.matching.api.EventPublisher;
import io.github.giovanicaprison.matching.api.MatchingEngine;
import io.github.giovanicaprison.matching.protocol.CancelOrderDecoder;
import io.github.giovanicaprison.matching.protocol.InstrumentDefinitionDecoder;
import io.github.giovanicaprison.matching.protocol.MassCancelDecoder;
import io.github.giovanicaprison.matching.protocol.MessageHeaderDecoder;
import io.github.giovanicaprison.matching.protocol.NewOrderDecoder;
import io.github.giovanicaprison.matching.protocol.PricingInstruction;
import io.github.giovanicaprison.matching.protocol.RejectReason;
import io.github.giovanicaprison.matching.protocol.RemoveReason;
import io.github.giovanicaprison.matching.protocol.ReplaceOrderDecoder;
import io.github.giovanicaprison.matching.protocol.SessionState;
import io.github.giovanicaprison.matching.protocol.SessionStateChangeDecoder;
import io.github.giovanicaprison.matching.protocol.TimeInForce;
import java.nio.ByteOrder;
import org.agrona.DirectBuffer;

/**
 * Rung three's lean twin: limit and market orders, price-time, on the flyweight layout.
 *
 * <p>Published matching engine figures are taken on books like this one, and METHODOLOGY's first
 * question is what the rest of a real venue's feature set costs when nobody uses it. A runtime flag
 * cannot answer that, because a disabled feature behind a branch still occupies the method, the
 * layout and the inlining budget (P-16). So this engine is the arm where the features do not exist:
 * no trigger book consulted after an execution, no tranche arithmetic on a take, no self match
 * comparison per candidate, no allocation choice, no auction, and an order slot that fits in one
 * cache line where the full rung's needs two. Comparing it against the full rung on flow that uses
 * only this remit measures the cost of existence itself.
 *
 * <p>On the shared remit this engine, the full flyweight rung, and every arm below them are byte
 * identical by construction, and tests hold them to it: an arm that answered differently would be
 * measuring a behaviour difference and calling it a feature cost.
 *
 * <p>Input is trusted the way every engine here trusts it (P-14). Flow at this composition never
 * carries a qualifier, so the fields this engine does not read are fields nothing sets.
 */
public final class LeanEngine implements MatchingEngine {

  private static final ByteOrder WIRE = ByteOrder.LITTLE_ENDIAN;

  private static final int HEADER_TEMPLATE = MessageHeaderDecoder.templateIdEncodingOffset();
  private static final int BODY = MessageHeaderDecoder.ENCODED_LENGTH;

  private static final int LIMIT = PricingInstruction.LIMIT.value();
  private static final int MARKET = PricingInstruction.MARKET.value();
  private static final int DAY = TimeInForce.DAY.value();

  private static final int PRE_OPEN = SessionState.PRE_OPEN.value();
  private static final int CONTINUOUS = SessionState.CONTINUOUS.value();
  private static final int CLOSED = SessionState.CLOSED.value();

  private static final int NON_POSITIVE_QUANTITY = RejectReason.NON_POSITIVE_QUANTITY.value();
  private static final int LOT_VIOLATION = RejectReason.LOT_VIOLATION.value();
  private static final int NON_POSITIVE_PRICE = RejectReason.NON_POSITIVE_PRICE.value();
  private static final int TICK_VIOLATION = RejectReason.TICK_VIOLATION.value();
  private static final int STATIC_BAND_VIOLATION = RejectReason.STATIC_BAND_VIOLATION.value();
  private static final int DYNAMIC_BAND_VIOLATION = RejectReason.DYNAMIC_BAND_VIOLATION.value();
  private static final int INVALID_FIELDS = RejectReason.INVALID_FIELDS.value();
  private static final int STATE_NOT_PERMITTED = RejectReason.STATE_NOT_PERMITTED.value();
  private static final int UNKNOWN_ORDER = RejectReason.UNKNOWN_ORDER.value();
  private static final int QUANTITY_BELOW_EXECUTED = RejectReason.QUANTITY_BELOW_EXECUTED.value();

  private static final int CANCELLED = RemoveReason.CANCELLED.value();
  private static final int REPLACED = RemoveReason.REPLACED.value();
  private static final int MASS_CANCELLED = RemoveReason.MASS_CANCELLED.value();
  private static final int IOC_REMAINDER = RemoveReason.IMMEDIATE_OR_CANCEL_REMAINDER.value();

  // Field offsets inside each command's body, each tied to its schema field by the decoder.
  private static final int DEFINITION_INSTRUMENT =
      InstrumentDefinitionDecoder.frameEncodingOffset();
  private static final int DEFINITION_TICK = InstrumentDefinitionDecoder.tickSizeEncodingOffset();
  private static final int DEFINITION_LOT = InstrumentDefinitionDecoder.lotSizeEncodingOffset();
  private static final int DEFINITION_MIN = InstrumentDefinitionDecoder.minPriceEncodingOffset();
  private static final int DEFINITION_MAX = InstrumentDefinitionDecoder.maxPriceEncodingOffset();
  private static final int DEFINITION_BAND = InstrumentDefinitionDecoder.bandWidthEncodingOffset();
  private static final int DEFINITION_OPEN =
      InstrumentDefinitionDecoder.openingReferenceEncodingOffset();

  private static final int NEW_CLIENT = NewOrderDecoder.clientOrderIdEncodingOffset();
  private static final int NEW_PARTICIPANT = NewOrderDecoder.participantIdEncodingOffset();
  private static final int NEW_SIDE = NewOrderDecoder.sideEncodingOffset();
  private static final int NEW_PRICING = NewOrderDecoder.pricingEncodingOffset();
  private static final int NEW_TIF = NewOrderDecoder.timeInForceEncodingOffset();
  private static final int NEW_PRICE = NewOrderDecoder.priceEncodingOffset();
  private static final int NEW_QUANTITY = NewOrderDecoder.quantityEncodingOffset();

  private static final int CANCEL_CLIENT = CancelOrderDecoder.clientOrderIdEncodingOffset();
  private static final int CANCEL_PARTICIPANT = CancelOrderDecoder.participantIdEncodingOffset();

  private static final int REPLACE_CLIENT = ReplaceOrderDecoder.clientOrderIdEncodingOffset();
  private static final int REPLACE_PARTICIPANT = ReplaceOrderDecoder.participantIdEncodingOffset();
  private static final int REPLACE_QUANTITY = ReplaceOrderDecoder.quantityEncodingOffset();
  private static final int REPLACE_PRICE = ReplaceOrderDecoder.priceEncodingOffset();

  private static final int MASS_CLIENT = MassCancelDecoder.clientOrderIdEncodingOffset();
  private static final int MASS_PARTICIPANT = MassCancelDecoder.participantIdEncodingOffset();

  private static final int SESSION_STATE = SessionStateChangeDecoder.stateEncodingOffset();

  private final Feed feed;
  private final Slab slab = new Slab(1 << 16);
  private final IntScratch gathered = new IntScratch();

  private Book book;

  private long tickSize = 1;
  private long lotSize = 1;
  private long minPrice;
  private long maxPrice;
  private long bandWidth;
  private long baseTick;
  private int state = PRE_OPEN;
  private long reference;
  private long nextOrderId = 1;
  private long nextExecutionId = 1;
  private long arrival;

  LeanEngine(final EventPublisher events) {
    this.feed = new Feed(events);
  }

  @Override
  public void onCommand(final DirectBuffer buffer, final int offset, final int length) {
    final int body = offset + BODY;
    switch (buffer.getShort(offset + HEADER_TEMPLATE, WIRE) & 0xFFFF) {
      case NewOrderDecoder.TEMPLATE_ID -> enter(buffer, body);
      case CancelOrderDecoder.TEMPLATE_ID -> cancel(buffer, body);
      case ReplaceOrderDecoder.TEMPLATE_ID -> replace(buffer, body);
      case MassCancelDecoder.TEMPLATE_ID -> massCancel(buffer, body);
      case SessionStateChangeDecoder.TEMPLATE_ID -> {
        state = buffer.getByte(body + SESSION_STATE);
        feed.stateChanged(state);
      }
      case InstrumentDefinitionDecoder.TEMPLATE_ID -> define(buffer, body);
      default ->
          throw new IllegalArgumentException(
              "template "
                  + (buffer.getShort(offset + HEADER_TEMPLATE, WIRE) & 0xFFFF)
                  + " is not a command (P-14)");
    }
  }

  private void define(final DirectBuffer buffer, final int body) {
    tickSize = buffer.getLong(body + DEFINITION_TICK, WIRE);
    lotSize = buffer.getLong(body + DEFINITION_LOT, WIRE);
    minPrice = buffer.getLong(body + DEFINITION_MIN, WIRE);
    maxPrice = buffer.getLong(body + DEFINITION_MAX, WIRE);
    bandWidth = buffer.getLong(body + DEFINITION_BAND, WIRE);
    reference = buffer.getLong(body + DEFINITION_OPEN, WIRE);
    baseTick = (minPrice + tickSize - 1) / tickSize;
    book = new Book(slab, tickSize, baseTick, (int) (maxPrice / tickSize - baseTick + 1));
    feed.instrument(buffer.getInt(body + DEFINITION_INSTRUMENT, WIRE));
  }

  // Order entry ---------------------------------------------------------------------------------

  private void enter(final DirectBuffer buffer, final int body) {
    final long clientOrderId = buffer.getLong(body + NEW_CLIENT, WIRE);
    final int participantId = buffer.getInt(body + NEW_PARTICIPANT, WIRE);
    final int side = buffer.getByte(body + NEW_SIDE);
    final int pricing = buffer.getByte(body + NEW_PRICING);
    final int timeInForce = buffer.getByte(body + NEW_TIF);
    final long price = buffer.getLong(body + NEW_PRICE, WIRE);
    final long quantity = buffer.getLong(body + NEW_QUANTITY, WIRE);

    final long verdict = refusalOrTick(pricing, timeInForce, price, quantity);
    if (verdict < 0) {
      feed.rejected(clientOrderId, participantId, reasonOf(verdict));
      return;
    }
    final long id = nextOrderId++;
    final int slot = slab.acquire();
    slab.init(
        slot,
        id,
        clientOrderId,
        participantId,
        side,
        pricing,
        timeInForce,
        (int) verdict,
        quantity,
        ++arrival,
        0);
    feed.accepted(id, clientOrderId, participantId);
    if (state == CONTINUOUS) {
      match(slot, side);
    }
    settle(slot, side);
  }

  private void settle(final int slot, final int side) {
    if (slab.remaining(slot) == 0) {
      slab.release(slot);
      return;
    }
    if (slab.pricing(slot) == LIMIT && slab.timeInForce(slot) <= DAY) {
      slab.rest(slot, ++arrival);
      book.add(side, slot);
      feed.rested(slab.id(slot), side, book.priceOfTick(slab.tick(slot)), slab.remaining(slot));
      return;
    }
    feed.removed(slab.id(slot), slab.remaining(slot), IOC_REMAINDER);
    slab.release(slot);
  }

  // Matching ------------------------------------------------------------------------------------

  /** (FR-3.1, FR-3.3) Best price first, then earliest arrival, until nothing crosses. */
  private void match(final int taker, final int side) {
    final int limitRank =
        slab.pricing(taker) == MARKET
            ? book.marketLimit()
            : book.rankOf(side ^ 1, slab.tick(taker));
    while (slab.remaining(taker) > 0) {
      final int resting = book.nextToTake(side, limitRank);
      if (resting == 0) {
        return;
      }
      final long quantity = Math.min(slab.remaining(taker), slab.remaining(resting));
      final long price = book.priceOfTick(slab.tick(resting));
      slab.take(taker, quantity);
      slab.take(resting, quantity);
      feed.executed(nextExecutionId++, slab.id(taker), slab.id(resting), price, quantity);
      reference = price;
      final int restingSide = side ^ 1;
      book.quantityChanged(restingSide, resting, -quantity);
      if (slab.remaining(resting) == 0) {
        book.remove(restingSide, resting);
        slab.release(resting);
      }
    }
  }

  // Amend and cancel ----------------------------------------------------------------------------

  private void cancel(final DirectBuffer buffer, final int body) {
    final long clientOrderId = buffer.getLong(body + CANCEL_CLIENT, WIRE);
    final int participantId = buffer.getInt(body + CANCEL_PARTICIPANT, WIRE);
    if (state == CLOSED) {
      feed.rejected(clientOrderId, participantId, STATE_NOT_PERMITTED);
      return;
    }
    final int resting = book.named(participantId, clientOrderId);
    if (resting == 0) {
      feed.rejected(clientOrderId, participantId, UNKNOWN_ORDER);
      return;
    }
    book.remove(slab.side(resting), resting);
    feed.removed(slab.id(resting), slab.remaining(resting), CANCELLED);
    slab.release(resting);
  }

  private void replace(final DirectBuffer buffer, final int body) {
    final long clientOrderId = buffer.getLong(body + REPLACE_CLIENT, WIRE);
    final int participantId = buffer.getInt(body + REPLACE_PARTICIPANT, WIRE);
    if (state == CLOSED) {
      feed.rejected(clientOrderId, participantId, STATE_NOT_PERMITTED);
      return;
    }
    final int resting = book.named(participantId, clientOrderId);
    if (resting == 0) {
      feed.rejected(clientOrderId, participantId, UNKNOWN_ORDER);
      return;
    }
    final long quantity = buffer.getLong(body + REPLACE_QUANTITY, WIRE);
    final long price = buffer.getLong(body + REPLACE_PRICE, WIRE);
    final long verdict = refusalOrTickForReplace(resting, quantity, price);
    if (verdict < 0) {
      feed.rejected(clientOrderId, participantId, reasonOf(verdict));
      return;
    }
    final int side = slab.side(resting);
    final int newTick = (int) verdict;
    final long remainder = quantity - slab.executed(resting);
    if (newTick == slab.tick(resting) && remainder < slab.remaining(resting)) {
      // (FR-4.4) Lowering quantity at the same price keeps queue position.
      final long remainingBefore = slab.remaining(resting);
      slab.reduceTo(resting, remainder);
      book.quantityChanged(side, resting, remainder - remainingBefore);
      feed.reduced(slab.id(resting), slab.remaining(resting));
      return;
    }
    book.remove(side, resting);
    feed.removed(slab.id(resting), slab.remaining(resting), REPLACED);
    // The same order under a new price or quantity, keeping both of its ids (FR-4.8).
    final long id = slab.id(resting);
    final int pricing = slab.pricing(resting);
    final int timeInForce = slab.timeInForce(resting);
    final long executed = slab.executed(resting);
    slab.release(resting);
    final int fresh = slab.acquire();
    slab.init(
        fresh,
        id,
        clientOrderId,
        participantId,
        side,
        pricing,
        timeInForce,
        newTick,
        remainder,
        ++arrival,
        executed);
    if (state == CONTINUOUS) {
      match(fresh, side);
    }
    settle(fresh, side);
  }

  private void massCancel(final DirectBuffer buffer, final int body) {
    final long clientOrderId = buffer.getLong(body + MASS_CLIENT, WIRE);
    final int participantId = buffer.getInt(body + MASS_PARTICIPANT, WIRE);
    if (state == CLOSED) {
      feed.rejected(clientOrderId, participantId, STATE_NOT_PERMITTED);
      return;
    }
    gathered.clear();
    book.of(participantId, gathered);
    gathered.sortByArrival(slab);
    for (int at = 0; at < gathered.size(); at++) {
      final int slot = gathered.get(at);
      book.remove(slab.side(slot), slot);
      feed.removed(slab.id(slot), slab.remaining(slot), MASS_CANCELLED);
      slab.release(slot);
    }
  }

  // Validation ----------------------------------------------------------------------------------

  /** A refusal encoded below zero, so one verdict carries either the reason or the tick. */
  private static long refusal(final int reason) {
    return -1 - (long) reason;
  }

  private static int reasonOf(final long verdict) {
    return (int) (-1 - verdict);
  }

  private long refusalOrTick(
      final int pricing, final int timeInForce, final long price, final long quantity) {
    if (state == CLOSED) {
      return refusal(STATE_NOT_PERMITTED);
    }
    if (quantity <= 0) {
      return refusal(NON_POSITIVE_QUANTITY);
    }
    if (lotSize != 1 && quantity % lotSize != 0) {
      return refusal(LOT_VIOLATION);
    }
    if (pricing == MARKET && timeInForce <= DAY) {
      // (VR-3.1) A market order cannot rest, so it cannot be told to.
      return refusal(INVALID_FIELDS);
    }
    if (pricing == LIMIT) {
      return tickOrRefusal(price);
    }
    return 0;
  }

  /**
   * The price's tick index, or the refusal that keeps it off the ladder, in one division: the
   * quotient that proves the price on tick (VR-2.2) is the index the ladder wants.
   */
  private long tickOrRefusal(final long price) {
    if (price <= 0) {
      return refusal(NON_POSITIVE_PRICE);
    }
    final long ticks = price / tickSize;
    if (ticks * tickSize != price) {
      return refusal(TICK_VIOLATION);
    }
    if (price < minPrice || price > maxPrice) {
      return refusal(STATIC_BAND_VIOLATION);
    }
    if (Math.abs(price - reference) > bandWidth) {
      return refusal(DYNAMIC_BAND_VIOLATION);
    }
    return ticks - baseTick;
  }

  private long refusalOrTickForReplace(final int resting, final long quantity, final long price) {
    if (quantity <= 0) {
      return refusal(NON_POSITIVE_QUANTITY);
    }
    if (quantity <= slab.executed(resting)) {
      return refusal(QUANTITY_BELOW_EXECUTED);
    }
    if (lotSize != 1 && quantity % lotSize != 0) {
      return refusal(LOT_VIOLATION);
    }
    return tickOrRefusal(price);
  }
}

File: cpp/lean-flyweight/lean_engine.cpp
#include "lean-flyweight/lean_engine.hpp"

#include <algorithm>
#include <cstdlib>
#include <stdexcept>
#include <string>

#include "io_github_giovanicaprison_matching_protocol/CancelOrder.h"
#include "io_github_giovanicaprison_matching_protocol/MassCancel.h"
#include "io_github_giovanicaprison_matching_protocol/MessageHeader.h"
#include "io_github_giovanicaprison_matching_protocol/PricingInstruction.h"
#include "io_github_giovanicaprison_matching_protocol/RejectReason.h"
#include "io_github_giovanicaprison_matching_protocol/RemoveReason.h"
#include "io_github_giovanicaprison_matching_protocol/ReplaceOrder.h"
#include "io_github_giovanicaprison_matching_protocol/SessionStateChange.h"
#include "io_github_giovanicaprison_matching_protocol/TimeInForce.h"

namespace io::github::giovanicaprison::matching::lean::flyweight {

namespace {

using protocol::MessageHeader;
using protocol::PricingInstruction;
using protocol::RejectReason;
using protocol::RemoveReason;
using protocol::SessionState;
using protocol::TimeInForce;

// The schema's numbers as plain ints, so side arithmetic and the verdict encoding read as the
// Java twin's do: BUY encodes to zero, so side ^ 1 is always the opposite side.
constexpr std::int32_t LIMIT = PricingInstruction::LIMIT;
constexpr std::int32_t MARKET = PricingInstruction::MARKET;
constexpr std::int32_t DAY = TimeInForce::DAY;
constexpr std::int32_t CONTINUOUS = SessionState::CONTINUOUS;
constexpr std::int32_t CLOSED = SessionState::CLOSED;

// A refusal encoded below zero, so one verdict carries either the reason or the tick.
constexpr std::int64_t refusal(const RejectReason::Value reason) {
  return -1 - static_cast<std::int64_t>(reason);
}

RejectReason::Value reasonOf(const std::int64_t verdict) {
  return static_cast<RejectReason::Value>(-1 - verdict);
}

template <typename Decoder>
Decoder decoded(char* buffer, const std::size_t body, const MessageHeader& header,
                const std::size_t end) {
  Decoder decoder;
  decoder.wrapForDecode(buffer, body, header.blockLength(), header.version(), end);
  return decoder;
}

}  // namespace

void LeanEngine::onCommand(char* buffer, const std::size_t offset, const std::size_t length) {
  MessageHeader header;
  const std::size_t end = offset + length;
  header.wrap(buffer, offset, 0, end);
  const std::size_t body = offset + MessageHeader::encodedLength();
  switch (header.templateId()) {
    case protocol::NewOrder::sbeTemplateId(): {
      auto newOrder = decoded<protocol::NewOrder>(buffer, body, header, end);
      enter(newOrder);
      return;
    }
    case protocol::CancelOrder::sbeTemplateId(): {
      auto command = decoded<protocol::CancelOrder>(buffer, body, header, end);
      cancel(command.clientOrderId(), command.participantId());
      return;
    }
    case protocol::ReplaceOrder::sbeTemplateId(): {
      auto command = decoded<protocol::ReplaceOrder>(buffer, body, header, end);
      replace(command.clientOrderId(), command.participantId(), command.quantity(),
              command.price());
      return;
    }
    case protocol::MassCancel::sbeTemplateId(): {
      auto command = decoded<protocol::MassCancel>(buffer, body, header, end);
      massCancel(command.clientOrderId(), command.participantId());
      return;
    }
    case protocol::SessionStateChange::sbeTemplateId(): {
      auto command = decoded<protocol::SessionStateChange>(buffer, body, header, end);
      state_ = command.state();
      feed_.stateChanged(static_cast<SessionState::Value>(state_));
      return;
    }
    case protocol::InstrumentDefinition::sbeTemplateId(): {
      auto definition = decoded<protocol::InstrumentDefinition>(buffer, body, header, end);
      define(definition);
      return;
    }
    default:
      throw std::invalid_argument("template " + std::to_string(header.templateId()) +
                                  " is not a command (P-14)");
  }
}

// The definition arrives once, before every other command (FR-1.1), which is what sizes the
// ladder.
void LeanEngine::define(protocol::InstrumentDefinition& definition) {
  tickSize_ = definition.tickSize();
  lotSize_ = definition.lotSize();
  minPrice_ = definition.minPrice();
  maxPrice_ = definition.maxPrice();
  bandWidth_ = definition.bandWidth();
  reference_ = definition.openingReference();
  baseTick_ = (minPrice_ + tickSize_ - 1) / tickSize_;
  book_.emplace(slab_, tickSize_, baseTick_,
                static_cast<std::int32_t>(maxPrice_ / tickSize_ - baseTick_ + 1));
  feed_.instrument(definition.frame().instrumentId());
}

// Order entry -------------------------------------------------------------------------------

void LeanEngine::enter(protocol::NewOrder& newOrder) {
  const std::uint64_t clientOrderId = newOrder.clientOrderId();
  const std::uint32_t participantId = newOrder.participantId();
  const std::int32_t side = newOrder.side();
  const std::int32_t pricing = newOrder.pricing();
  const std::int32_t timeInForce = newOrder.timeInForce();
  const std::int64_t price = newOrder.price();
  const std::int64_t quantity = newOrder.quantity();

  const std::int64_t verdict = refusalOrTick(pricing, timeInForce, price, quantity);
  if (verdict < 0) {
    feed_.rejected(clientOrderId, participantId, reasonOf(verdict));
    return;
  }
  const std::uint64_t id = nextOrderId_++;
  const std::int32_t slot = slab_.acquire();
  slab_.init(slot, id, clientOrderId, participantId, side, pricing, timeInForce,
             static_cast<std::int32_t>(verdict), quantity, ++arrival_, 0);
  feed_.accepted(id, clientOrderId, participantId);
  if (state_ == CONTINUOUS) {
    match(slot, side);
  }
  settle(slot, side);
}

void LeanEngine::settle(const std::int32_t slot, const std::int32_t side) {
  if (slab_.remaining(slot) == 0) {
    slab_.release(slot);
    return;
  }
  if (slab_.pricing(slot) == LIMIT && slab_.timeInForce(slot) <= DAY) {
    slab_.rest(slot, ++arrival_);
    book_->add(side, slot);
    feed_.rested(slab_.id(slot), side, book_->priceOfTick(slab_.tick(slot)), slab_.remaining(slot));
    return;
  }
  feed_.removed(slab_.id(slot), slab_.remaining(slot), RemoveReason::IMMEDIATE_OR_CANCEL_REMAINDER);
  slab_.release(slot);
}

// Matching ----------------------------------------------------------------------------------

// (FR-3.1, FR-3.3) Best price first, then earliest arrival, until nothing crosses.
void LeanEngine::match(const std::int32_t taker, const std::int32_t side) {
  const std::int32_t limitRank = slab_.pricing(taker) == MARKET
                                     ? book_->marketLimit()
                                     : book_->rankOf(side ^ 1, slab_.tick(taker));
  while (slab_.remaining(taker) > 0) {
    const std::int32_t resting = book_->nextToTake(side, limitRank);
    if (resting == 0) {
      return;
    }
    const std::int64_t quantity = std::min(slab_.remaining(taker), slab_.remaining(resting));
    const std::int64_t price = book_->priceOfTick(slab_.tick(resting));
    slab_.take(taker, quantity);
    slab_.take(resting, quantity);
    feed_.executed(nextExecutionId_++, slab_.id(taker), slab_.id(resting), price, quantity);
    reference_ = price;
    const std::int32_t restingSide = side ^ 1;
    book_->quantityChanged(restingSide, resting, -quantity);
    if (slab_.remaining(resting) == 0) {
      book_->remove(restingSide, resting);
      slab_.release(resting);
    }
  }
}

// Amend and cancel --------------------------------------------------------------------------

void LeanEngine::cancel(const std::uint64_t clientOrderId, const std::uint32_t participantId) {
  if (state_ == CLOSED) {
    feed_.rejected(clientOrderId, participantId, RejectReason::STATE_NOT_PERMITTED);
    return;
  }
  const std::int32_t resting = book_->named(participantId, clientOrderId);
  if (resting == 0) {
    feed_.rejected(clientOrderId, participantId, RejectReason::UNKNOWN_ORDER);
    return;
  }
  book_->remove(slab_.side(resting), resting);
  feed_.removed(slab_.id(resting), slab_.remaining(resting), RemoveReason::CANCELLED);
  slab_.release(resting);
}

void LeanEngine::replace(const std::uint64_t clientOrderId, const std::uint32_t participantId,
                         const std::int64_t quantity, const std::int64_t price) {
  if (state_ == CLOSED) {
    feed_.rejected(clientOrderId, participantId, RejectReason::STATE_NOT_PERMITTED);
    return;
  }
  const std::int32_t resting = book_->named(participantId, clientOrderId);
  if (resting == 0) {
    feed_.rejected(clientOrderId, participantId, RejectReason::UNKNOWN_ORDER);
    return;
  }
  const std::int64_t verdict = refusalOrTickForReplace(resting, quantity, price);
  if (verdict < 0) {
    feed_.rejected(clientOrderId, participantId, reasonOf(verdict));
    return;
  }
  const std::int32_t side = slab_.side(resting);
  const std::int32_t newTick = static_cast<std::int32_t>(verdict);
  const std::int64_t remainder = quantity - slab_.executed(resting);
  if (newTick == slab_.tick(resting) && remainder < slab_.remaining(resting)) {
    // (FR-4.4) Lowering quantity at the same price keeps queue position.
    const std::int64_t remainingBefore = slab_.remaining(resting);
    slab_.reduceTo(resting, remainder);
    book_->quantityChanged(side, resting, remainder - remainingBefore);
    feed_.reduced(slab_.id(resting), slab_.remaining(resting));
    return;
  }
  // (FR-4.5) Anything else is a removal and a fresh rest, keeping both ids (FR-4.8).
  book_->remove(side, resting);
  feed_.removed(slab_.id(resting), slab_.remaining(resting), RemoveReason::REPLACED);
  const std::uint64_t id = slab_.id(resting);
  const std::int32_t pricing = slab_.pricing(resting);
  const std::int32_t timeInForce = slab_.timeInForce(resting);
  const std::int64_t executed = slab_.executed(resting);
  slab_.release(resting);
  const std::int32_t fresh = slab_.acquire();
  slab_.init(fresh, id, clientOrderId, participantId, side, pricing, timeInForce, newTick,
             remainder, ++arrival_, executed);
  if (state_ == CONTINUOUS) {
    match(fresh, side);
  }
  settle(fresh, side);
}

// (FR-4.7) Everything for one participant, in arrival order.
void LeanEngine::massCancel(const std::uint64_t clientOrderId, const std::uint32_t participantId) {
  if (state_ == CLOSED) {
    feed_.rejected(clientOrderId, participantId, RejectReason::STATE_NOT_PERMITTED);
    return;
  }
  gathered_.clear();
  book_->of(participantId, gathered_);
  sortByArrival(gathered_);
  for (const std::int32_t slot : gathered_) {
    book_->remove(slab_.side(slot), slot);
    feed_.removed(slab_.id(slot), slab_.remaining(slot), RemoveReason::MASS_CANCELLED);
    slab_.release(slot);
  }
}

// Earliest first, which is every tie-break and every report order in the venue.
void LeanEngine::sortByArrival(std::vector<std::int32_t>& slots) const {
  std::sort(slots.begin(), slots.end(), [this](const std::int32_t left, const std::int32_t right) {
    return slab_.arrival(left) < slab_.arrival(right);
  });
}

// Validation --------------------------------------------------------------------------------

std::int64_t LeanEngine::refusalOrTick(const std::int32_t pricing, const std::int32_t timeInForce,
                                       const std::int64_t price,
                                       const std::int64_t quantity) const {
  if (state_ == CLOSED) {
    return refusal(RejectReason::STATE_NOT_PERMITTED);
  }
  if (quantity <= 0) {
    return refusal(RejectReason::NON_POSITIVE_QUANTITY);
  }
  if (lotSize_ != 1 && quantity % lotSize_ != 0) {
    return refusal(RejectReason::LOT_VIOLATION);
  }
  if (pricing == MARKET && timeInForce <= DAY) {
    // (VR-3.1) A market order cannot rest, so it cannot be told to.
    return refusal(RejectReason::INVALID_FIELDS);
  }
  if (pricing == LIMIT) {
    return tickOrRefusal(price);
  }
  return 0;
}

// The price's tick index, or the refusal that keeps it off the ladder, in one division: the
// quotient that proves the price on tick (VR-2.2) is the index the ladder wants.
std::int64_t LeanEngine::tickOrRefusal(const std::int64_t price) const {
  if (price <= 0) {
    return refusal(RejectReason::NON_POSITIVE_PRICE);
  }
  const std::int64_t ticks = price / tickSize_;
  if (ticks * tickSize_ != price) {
    return refusal(RejectReason::TICK_VIOLATION);
  }
  if (price < minPrice_ || price > maxPrice_) {
    return refusal(RejectReason::STATIC_BAND_VIOLATION);
  }
  if (std::abs(price - reference_) > bandWidth_) {
    return refusal(RejectReason::DYNAMIC_BAND_VIOLATION);
  }
  return ticks - baseTick_;
}

std::int64_t LeanEngine::refusalOrTickForReplace(const std::int32_t resting,
                                                 const std::int64_t quantity,
                                                 const std::int64_t price) const {
  if (quantity <= 0) {
    return refusal(RejectReason::NON_POSITIVE_QUANTITY);
  }
  if (quantity <= slab_.executed(resting)) {
    // (FR-4.9) Nothing can un-trade what has traded.
    return refusal(RejectReason::QUANTITY_BELOW_EXECUTED);
  }
  if (lotSize_ != 1 && quantity % lotSize_ != 0) {
    return refusal(RejectReason::LOT_VIOLATION);
  }
  return tickOrRefusal(price);
}

}  // namespace io::github::giovanicaprison::matching::lean::flyweight
